### PR TITLE
[JENKINS-61973] Restyle sidebar tasks

### PR DESCRIPTION
**Pull request by fqueiruga**
See [JENKINS-61973](https://issues.jenkins-ci.org/browse/JENKINS-61973).

Revamp of sidebar task list

- Refactored and simplified the <l:task> tag. Tasks now only have 1 hyperlink that contains both labels and icon.
- Accessibility of the task list was improved: keyboard navigation is easier now and interactive states have better visual cues.

<details>
<summary>Screenshots</summary>

Homepage tasks
Captura de pantalla 2020-05-05 a las 17 26 41

Example of disabled state
Captura de pantalla 2020-04-30 a las 13 02 22

Active page, focus and hover states
Captura de pantalla 2020-04-30 a las 13 02 56

Hover states and on a nested page
Captura de pantalla 2020-05-05 a las 17 27 21

Full page:
Captura de pantalla 2020-05-05 a las 17 12 01

Full page 2:
Captura de pantalla 2020-05-05 a las 17 12 13

</details>

### Proposed changelog entries

* Revamped the sidebar task list, with improved aesthetics and accessibility.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@timja
@uhafner
@daniel-beck 
@Wadeck 
@oleg-nenashev 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Comment by daniel-beck** [comment]
The nested UI looks kind of weird IMO:

> Screenshot 2020-04-30 at 14 40 38

> Screenshot 2020-04-30 at 14 40 49

**Comment by daniel-beck** [comment]
Would it make sense to adapt the context menus in a similar manner, at least e.g. the hover states?

File: core/src/main/resources/lib/layout/task.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/task.jelly:88]: what's this newDomain check for?

Comment (fqueiruga) [reply core/src/main/resources/lib/layout/task.jelly]: oops I think that's dev code that I forgot to remove, it should be fixed now

**Review by timja — APPROVED** [review]
Tested locally, looks good to me

**Comment by uhafner** [comment]
Combining icon and text looks much better now! Also the spacing is better as before: it provides a more "clean" look (I'm not sure if this is the right English word). 

When I start Jenkins it looks a little bit weird, because the blue vertical line does not match well with the black horizontal Jenkins border. Maybe we should start to make the new UI the default as well? With the blue Jenkins border it matches perfectly. Otherwise you would need to make the vertical line black for that theme. 

What could be improved in my opinion: the highlighting (on hover and after selection) looks a little bit strange since the rectangle ends somewhere in the middle of the screen:

Bildschirmfoto 2020-05-01 um 18 25 53

It would make sense to have a horizontal line (or background?) that indicates the whole area of the left area. (E.g. like in Gitter or in any other chat tool: these tools use a different color for the navigation bar than the rest of the UI.)

Example:

Bildschirmfoto 2020-05-01 um 18 28 51

Also the navigation bar and the breadcrumbs seem not to "belong together" with respect to the UI styles. Somehow one gets the feeling that these things are not designed using the same color scheme.

Bildschirmfoto 2020-05-01 um 18 19 04

And a minor thing: the new styling seems not to be consistent with the new system configuration hovering: there the background is gray with an border, in the navigation bar we have only gray and no border.

Then I have two additional topics, that might make sense to be discussed separately (but we should already think about them during this PR):
- it would be helpful, if the navigation bar would be fixed and is not scrollable anymore 
- When you already change the jelly tags: it would make sense to add an option to provide SVG icons and badges (e.g. to show the number of failed tests, number of warnings, etc.).

File: war/src/main/less/modules/sidebar-tasks.less
Comment (varyvol) [inline war/src/main/less/modules/sidebar-tasks.less:56]: Shouldn't this be in the colors file?

File: war/src/main/less/modules/sidebar-tasks.less
Comment (varyvol) [inline war/src/main/less/modules/sidebar-tasks.less:118]: Ditto.

File: war/src/main/less/modules/sidebar-tasks.less
Comment (fqueiruga) [reply war/src/main/less/modules/sidebar-tasks.less:56]: TBH yes, but I have yet to figure out how all greys fill in the colour palette. I'm waiting until I have a better idea of the greyscale before extracting them to variables. I can extract them now if you think it's needed.

File: war/src/main/less/modules/sidebar-tasks.less
Comment (fqueiruga) [reply war/src/main/less/modules/sidebar-tasks.less:118]: See above

File: core/src/main/resources/lib/layout/task.jelly
Comment (fqueiruga) [reply core/src/main/resources/lib/layout/task.jelly:154]: I'd rather keep this PR just to UI improvements. If you think it's needed maybe @Wadeck can advise how to go about it.

File: war/src/main/less/modules/sidebar-tasks.less
Comment (varyvol) [reply war/src/main/less/modules/sidebar-tasks.less:56]: No problem, extract them when you have the full idea.

File: war/src/main/less/modules/sidebar-tasks.less
Comment (fqueiruga) [reply war/src/main/less/modules/sidebar-tasks.less]: That was more of a reminder that I forgot to remove. The max task width is limited to 325px and I'm not sure what caused that limitation.

File: war/src/main/less/base-styles-v2.less
Comment (fqueiruga) [reply war/src/main/less/base-styles-v2.less]: Makes sense, done

**Comment by fqueiruga** [comment]
Adressing your feedback Ulli:

- I don't think the blue vertical line looks off with a black header. It's normal (and a good thing IMO) to have splashes of colour. It's the same case as the primary buttons.
- I'll try to play with task link width a bit with some options and I'll get back to you. I agree that the sidepanel should be of a different colour than the main part (example [1]), but it requires deeper changes also on layout level. I think it's a bit out of scope for this, something to definitely consider in the future.
- I don't think the breadcrumbs look off. Especially if we keep incorporating these new greys and black hyperlinks (as we will probably do when working on tables & tabs). Regarding the header, I think it's a candidate for revisiting. It was the first component we worked on, it's not perfect.
- I don't think it's a problem that the hover is a bit different than the one on /manage. Dropdown active states definitely look weird but they are complex to improve (some weird YUI black magic going on). Something to improve in the future definitely.

I'll relay these feedback items to @josephbrueggen and get back to you. I'll play around with hover states a bit to see if they can be better.

Regarding your other conversation starters, I think fixing the navbar can be a bit too much,as  we already have sticky breadcrumbs. Allowing badges seems a great idea for the future. Maybe we can discuss it in the next meeting.

<details>
<summary>[1] sidebar with different colour</summary>

Captura de pantalla 2020-05-04 a las 11 42 18

</details>

**Comment by fqueiruga** [comment]
@uhafner I have updated the PR adding a border to the side panel. I needed to do some deep changes to the layout-commons file in order to have the side-panel to stretch all the way to the bottom. If these changes are better for a separate PR I'll happily roll them back, and create a separate PR.

<details>
<summary>New screenshots (needs some padding on the main-panel IMO)</summary>

Captura de pantalla 2020-05-05 a las 17 12 01

Captura de pantalla 2020-05-05 a las 17 12 13

Side effects: some pages use an empty side-panel:
Captura de pantalla 2020-05-05 a las 17 10 33

</details>

**Comment by fqueiruga** [comment]
Re-requesting reviews as I have updated the side-panel to have a border (see comments above)

**Comment by josephbrueggen** [comment]
@uhafner thank you for the great feedback! I agree with @fqueiruga's responses above. I also wanted to add the following:

- Changing the background color of the side panel is an option that I'm exploring. It's beyond scope for now, but rest assured I am mindful of this idea and we may pursue it in the future as we continue to evolve this UI.
- You've made an excellent point about the side panel benefitting from a visible edge. This would help delineate between areas of the interface, and it would help make the overall interface more intuitive. See Felix's comment above.
- I agree with you that the breadcrumbs warrant some slight visual adjustments to better align with the side panel. As the design system becomes more well-rounded, we will occasionally need to revisit components in this way. Let's discuss this in the next SIG meeting, too.

**Comment by uhafner** [comment]
It looks really nice with the new separation line!

**Comment by daniel-beck** [comment]
> Side effects: some pages use an empty side-panel

Highlights pages that probably should just be using `one-column` layout (see `layout.jelly`).

**Comment by daniel-beck** [comment]
> New screenshots (needs some padding on the main-panel IMO)

I wonder whether we can make the exact opposite work, e.g. let the builds list table of the list view (or plugin manager table) span the entire width.

**Comment by fqueiruga** [comment]
> I wonder whether we can make the exact opposite work, e.g. let the builds list table of the list view (or plugin manager table) span the entire width.

That's sort of the plan and what the follow up on side panel work is about (https://issues.jenkins-ci.org/browse/JENKINS-62175).

<details>
<summary>Mock by @josephbrueggen </summary>

Captura de pantalla 2020-05-06 a las 12 24 59

</details

**Comment by daniel-beck** [comment]
@fqueiruga To clarify, I meant the main panel, rather than the side panel:


<details>
<summary>mock</summary>

81083818-54db2080-8ef5-11ea-813e-952947a77656


</details>

**Comment by fqueiruga** [comment]
I think that won't look good and will not scale on any other screen. We need to be mindful of other pages on the main panel (forms, plugins, other tables, etc)

**Comment by daniel-beck** [comment]
> on any other screen

If it's just that it can be a property of the main panel whether to have padding or not. Then views like this can be optimized for it, while others get the default padding.

Obviously if it doesn't look good then we don't need it anyway 😄

**Comment by fqueiruga** [comment]
TBH you may be onto something. We are still going to look into tables & tabs rather soon so maybe it's best to wait.

I like the idea of enhancing the #main-panel: setting max-width & center the content, etc. But I think we need to work on some stuff first :D

**Review by varyvol — APPROVED** [review]

**Comment by batmat** [comment]
@timja I see you originally approved, and have been requested for review again, can you please have another look and approve again if fine (or not, if not :-)). Thanks

**Review by timja — APPROVED** [review]
I haven't checked after the tweaks, code changes are fine

**Comment by batmat** [comment]
@uhafner it seems like your feedback was addressed by Félix, and your last comment would imply you're fine with this change now? Could you please approve it if so, or tell us what you're missing to have this acceptable? Danke schön!

**Review by uhafner — APPROVED** [review]
👍 

(I did not look into the code, just for the UI)

**Review by alecharp — APPROVED** [review]
I tested it locally with `docker run --rm -ti -p 8080:8080 -e ID=4700 jenkins/core-pr-tester` and it's working great! Couldn't spot any issue. 

Thank you!

**Comment by fqueiruga** [comment]
@oleg-nenashev @daniel-beck @Wadeck is there anything blocking this one for you? If this is going in I think it should be before the hackathon.

**Comment by oleg-nenashev** [comment]
I plan to merge it in 24 hours if no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information

**Pull request merged**
